compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp

```diff
@@ -15,6 +15,138 @@ namespace mlir::iree_compiler::AMDAIE {
+    int64_t sizeAfterSplit =
+        std::accumulate(memrefShape.begin() + objFifoSplitDim + 1,
+                        memrefShape.end(), 1, std::multiplies<>());
+
+    size_t sourceSplitDim{0};
+    size_t targetSplitDim{0};
+    if (dmaOp.getTargetObjectFifo() == objFifo) {
+      std::optional<SmallVector<int64_t>> targetSizes =
+          getConstantIntValues(dmaOp.getTargetMixedSizes());
+      std::optional<SmallVector<int64_t>> targetStrides =
+          getConstantIntValues(dmaOp.getTargetMixedStrides());
+      std::optional<SmallVector<int64_t>> sourceSizes =
+          getConstantIntValues(dmaOp.getSourceMixedSizes());
+      if (!targetSizes.has_value() || !targetStrides.has_value() ||
+          !sourceSizes.has_value()) {
+        return dmaOp.emitOpError() << "has unsupported dynamic target strides "
+                                      "or sizes or source sizes";
+      }
+      for (auto iter : llvm::enumerate(
+               llvm::zip(targetSizes.value(), targetStrides.value()))) {
+        int64_t size = std::get<0>(iter.value());
+        int64_t stride = std::get<1>(iter.value());
+        if (stride == sizeAfterSplit && size != 1) {
+          targetSplitDim = iter.index();
+          break;
+        }
+      }
+      int64_t targetSizeAfterSplit =
+          std::accumulate(targetSizes.value().begin() + targetSplitDim + 1,
+                          targetSizes.value().end(), 1, std::multiplies<>());
+      SmallVector<int64_t> sourceProductSizes = sourceSizes.value();
+      std::partial_sum(sourceProductSizes.rbegin(), sourceProductSizes.rend(),
+                       sourceProductSizes.rbegin(), std::multiplies<int64_t>());
+      for (int idx = sourceProductSizes.size() - 1; idx > 0; idx--) {
+        if (sourceProductSizes[idx] > targetSizeAfterSplit) {
+          sourceSplitDim = idx;
+          break;
+        }
+      }
+    } else if (dmaOp.getSourceObjectFifo() == objFifo) {
+      // Find outermost dimension in the access pattern that has stride ==
+      // sizeAfterSplit and size != 1.
+      std::optional<SmallVector<int64_t>> sourceSizes =
+          getConstantIntValues(dmaOp.getSourceMixedSizes());
+      std::optional<SmallVector<int64_t>> sourceStrides =
+          getConstantIntValues(dmaOp.getSourceMixedStrides());
+      std::optional<SmallVector<int64_t>> targetSizes =
+          getConstantIntValues(dmaOp.getTargetMixedSizes());
+      if (!sourceSizes.has_value() || !sourceStrides.has_value() ||
+          !targetSizes.has_value()) {
+        return dmaOp.emitOpError() << "has unsupported dynamic source strides "
+                                      "or sizes or target sizes";
+      }
+      for (auto iter : llvm::enumerate(
+               llvm::zip(sourceSizes.value(), sourceStrides.value()))) {
+        int64_t size = std::get<0>(iter.value());
+        int64_t stride = std::get<1>(iter.value());
+        if (stride == sizeAfterSplit && size != 1) {
```

Comment:
Looks like these codes are mostly duplicated by just switching source/target. Might be good to simplify these.